pkg/cluster/operation/upgrade.go

```diff
@@ -142,3 +179,86 @@ func Addr(ins spec.Instance) string {
 	}
 	return ins.GetHost() + ":" + strconv.Itoa(ins.GetPort())
 }
+
+var (
+	leaderScheduleLimitOffset = 32
+	regionScheduleLimitOffset = 512
+	//storeLimitOffset             = 512
+	leaderScheduleLimitThreshold = 64
+	regionScheduleLimitThreshold = 1024
+	//storeLimitThreshold          = 1024
+)
+
+// increaseScheduleLimit increases the schedule limit of leader and region for faster
+// rebalancing during the rolling restart / upgrade process
+func increaseScheduleLimit(ctx context.Context, pc *api.PDClient) (
+	currLeaderScheduleLimit int,
+	currRegionScheduleLimit int,
+	err error) {
+	// insert checkpoint
+	point := checkpoint.Acquire(ctx, map[string]interface{}{
+		"operation": "upgrade",
+		"func":      "increaseScheduleLimit",
+	})
+	defer func() {
+		point.Release(err,
+			zap.String("operation", "upgrade"),
+			zap.String("func", "increaseScheduleLimit"),
+			zap.Int("currLeaderScheduleLimit", currLeaderScheduleLimit),
+			zap.Int("currRegionScheduleLimit", currRegionScheduleLimit),
+		)
+	}()
+
+	if data := point.Hit(); data != nil {
+		currLeaderScheduleLimit = data["currLeaderScheduleLimit"].(int)
```

Comment:
I think this may panic because `currLeaderScheduleLimit` and `currRegionScheduleLimit` will be stored as float64 in json.